pycytominer/operations/transform.py

```diff
@@ -67,58 +77,94 @@ def fit(self, X, y=None):
         self
             With computed weights attribute
         """
-        # Get the mean of the features (columns) and center if specified
-        self.mu = X.mean()
-        if self.center:
-            X = X - self.mu
-
-        # Get the covariance matrix
-        C = (1 / X.shape[0]) * np.dot(X.transpose(), X)
+        # Get Numpy representation of the DataFrame
+        X = X.values
 
-        if self.method in ["PCA", "ZCA"]:
-            # Get the eigenvalues and eigenvectors of the covariance matrix
-            s, U = eigh(C)
+        if self.method in ["PCA-cor", "ZCA-cor"]:
+            # The projection matrix for PCA-cor and ZCA-cor is the same as the
+            # projection matrix for PCA and ZCA, respectively, on the standardized
+            # data. So, we first standardize the data, then compute the projection
 
-            # Fix sign ambiguity of eigenvectors
-            U = pd.DataFrame(U * np.sign(np.diag(U)))
+            self.standard_scaler = StandardScaler().fit(X)
+            variances = self.standard_scaler.var_
+            if np.any(variances == 0):
+                raise ValueError(
+                    "Divide by zero error, make sure low variance columns are removed"
+                )
 
-            # Process the eigenvalues into a diagonal matrix and fix rounding errors
-            D = np.diag(1.0 / np.sqrt(s.clip(self.epsilon)))
+            X = self.standard_scaler.transform(X)
+        else:
+            if self.center:
+                self.mean_centerer = StandardScaler(with_mean=True, with_std=False).fit(
+                    X
+                )
+                X = self.mean_centerer.transform(X)
 
-            # Calculate the sphering matrix
-            self.W = np.dot(D, U.transpose())
+        # Get the number of observations and variables
+        n, d = X.shape
 
-            # If ZCA, perform additional rotation
-            if self.method == "ZCA":
-                self.W = np.dot(U, self.W)
+        # compute the rank of the matrix X
+        r = np.linalg.matrix_rank(X)
 
-        if self.method in ["PCA-cor", "ZCA-cor"]:
-            # Get the correlation matrix
-            R = np.corrcoef(X.transpose())
+        # If n < d, then rank should be equal to n - 1 (if centered) or n (if not centered)
+        # If n >= d, then rank should be equal to d
+        if not ((r == d) | (self.center & (r == n - 1)) | (not self.center & (r == n))):
+            raise ValueError(
+                (
+                    "Sphering is not supported when the data matrix X is not full rank."
```

Comment:
Consider mentioning why sphering is not supported in this case.

Reply:
Great catch!

Sphering is indeed supported in this case, but it's just highly unlikely that we'd end up with a perfect linear dependency, so we should not proceed. I have updated the error message. LMK if this suffices.

Reply:
This looks like it will suffice!